### PR TITLE
Wint8 gemm and gemv opt

### PR DESCRIPTION
**Pull request by wwbitejotunn**
### PR types
Performance optimization

### PR changes
OPs

### Description
Pcard-71501

This PR optimize the speed of weight only gemm and gemv gpu kernel.

To speed up the weight-only gemm, following features were adopted
- using stream-k gemm instead of serial split-k
- using multi-warp batch gemv
- gemm/gemv dispatch based on problem size m

For gemms with problem sizes in llama13b, we obtain a 1.34x gemm kernel speed in A100 80G.

image

### CODE REVIEW

**Comment by vivienfanghuagood** [comment]
私以为，升级点在PR里可能应该更详细一点～

**Review by MARD1NO — APPROVED** [review]
Genius！

File: paddle/phi/infermeta/unary.cc [hunk 1]
```diff
@@ -5163,7 +5163,7 @@ void WeightQuantizeInferMeta(const MetaTensor& x,
   out->set_dtype(DataType::INT8);
 
   scale->set_dims(phi::make_ddim(dim_scale));
-  scale->set_dtype(DataType::FLOAT32);
```
Comment (zhoutianzi666) [inline paddle/phi/infermeta/unary.cc:5225]: 以后scale都是fp16了？

**Review by zhoutianzi666 — DISMISSED** [review]

File: paddle/phi/infermeta/unary.cc (hunk 1 above)
Comment (wwbitejotunn) [reply paddle/phi/infermeta/unary.cc:5225]: 对的, 这边scale改为了bf16/fp16, 能够有更好的性能, 精度应该也能保证, 已经同步修改了weight_quant op中计算scale的部分, 以及paddlenlp代码中scale权重初始化的逻辑

**Comment by wwbitejotunn** [comment]
> 私以为，升级点在PR里可能应该更详细一点～

已添加了升级点和测试数据~

**Review by heavengate — DISMISSED** [review]

**Review by raindrops2sea — DISMISSED** [review]

**Review by Xreki — DISMISSED** [review]
LGTM for const_cast

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**